### ISSUE TITLE
Public API for `installDependencies` and `resolveDepsAndWritePackageJson`: Distinguishing `installDependencies` tasks from `postSynthesize`

### ISSUE THREAD

**Issue by moltar** [projen/projen#2890]
I have a use case where I'd like to be able to distinguish the `installDependencies` processes from the `postSynthesize` task.

1. First-time `installDependencies` runs it installs using the `*` version(s).
2. These versions get updated by `resolveDepsAndWritePackageJson`.
3. Finally `installDependencies` runs again to update the lock file with the newly updated versions.

Where I need to be able to tell these steps apart is when using `syncpack` in a monorepo environment. `syncpack` needs to run _after_ `resolveDepsAndWritePackageJson` finishes to sync the versions.

It cannot run when the versions are `*` (aka latest), as it will sync those across the project because this is the highest possible version as far as SemVer is concerned.

Even just having an environment variable set would be enough for my purposes. But having a separate task would be the best solution.

I feel like this can also be related to #2889.

---

https://github.com/projen/projen/blob/5780c30daddbd8247bd3d64c2d6da69760d6533e/src/javascript/node-package.ts#L840-L857

**Comment by moltar** [comment]
> I recommend you override

This is what I do now.

But there's still no way to distinguish between these two calls, as there are no params to the method.

I could, in theory, use an invocation counter, and consider the first invocation to be the `*` install, and the second to be the versioned install. But then it's just turtles all the way down 😁

**Comment by moltar** [comment]
@mrgrain I noticed you renamed this to add "Public API ..."

Is the task named `install` considered to be private?

This is the way I was hooking into it using the public API methods:

```ts
project.tasks.tryFind('install')?.exec('foo')
```

**Comment by mrgrain** [comment]
The task is public. The methods you are referring to are not.

**Comment by moltar** [comment]
Right! I guess I was describing it in terms of methods for clarity.

But really, it could be solved by having separate tasks for the initial install with `*` version, and then subsequent install with filled versions. Even if under the hood those tasks would execute the same command, then it would open up this possibility.

**Comment by mrgrain** [comment]
Yeah, absolutely. I totally agree with the need for a public, supported API here. I just don't want to preclude the solution for this. I also kind of think the current implementation isn't great and probably should not be too exposed to users.
